src/python/WMCore/MicroService/MSCore/MSAuth.py

```diff
@@ -0,0 +1,115 @@
+where
+- role defines authorization role, e.g. admin
+- group defines specific group, e.g. reqmgr
+- service identifies MS service, e.g. ms-pileup
+- actions refers to a specific API or action, e.g. create
+- method contains list of allowed HTTP methods, e.g. ["POST", "PUT"]
+
+For example, here is a possible list of authz rules for MSPileup service
+
+authz_rules = \
+    [{"role": "admin", "group": "reqmgr", "service": "ms-pileup", "action": "create", "method": ["POST"]},
+     {"role": "production-operator", "group": "dataops", "service": "ms-pileup", "action": "create", "method": ["POST"]},
+
+     {"role": "admin", "group": "reqmgr", "service": "ms-pileup", "action": "create", "method": ["DELETE"]},
+     {"role": "production-operator", "group": "dataops", "service": "ms-pileup", "action": "create", "method": ["DELETE"]},
+
+     {"role": "admin", "group": "reqmgr", "service": "ms-pileup", "action": "update", "method": ["PUT"]},
+     {"role": "production-operator", "group": "dataops", "service": "ms-pileup", "action": "update", "method": ["PUT"]}]
+"""
+
+# system modules
+import os
+import json
+
+# third party modules
+import cherrypy
+
+# WMCore modules
+from WMCore.MicroService.Tools.Common import getMSLogger
+from WMCore.REST.Auth import user_info_from_headers
+
+
+def readAuthzRules(entry):
+    """
+    Return authz rules from given entry
+    :param entry: either file name or list of authz rules
+    :return: list of authz rules
+    """
+    rules = []
+    if entry is None:
+        return rules
+    # of provided is a list data-type it may contaisn our rules
+    if isinstance(entry, list) and len(entry) > 0:
+        if isinstance(entry[0], dict):
+            return entry
+    # check if entry is a file name
+    if os.path.exists(entry):
+        with open(entry, 'r', encoding="utf-8") as fstream:
+            rules = json.load(fstream)
+            return rules
+    msg = f"Unable to read {entry} as authz rules"
+    raise Exception(msg)
+
+
+class MSAuth():
+    """
+    This class provides auth/authz functionality for micro-services
+    """
+
+    def __init__(self, msConfig, **kwargs):
+        """
+        Provides a basic setup for all the microservices
+
+        :param config: MS service configuration
+        :param kwargs: optional parameters
+        """
+        self.logger = getMSLogger(getattr(msConfig, 'verbose', False), kwargs.get("logger"))
+        self.msConfig = msConfig
+        self.logger.info("Configuration including default values:\n%s", self.msConfig)
+
+        # load auth/authz configuration
+        self.authzRules = readAuthzRules(msConfig.get('authz_rules', None))
+        self.authzKey = msConfig.get('authz_key', None)
+
+    def authorizeApiAccess(self, service, action, method=None):
+        """
+        Check auth role.
+        :return: boolean
+        """
+        # get user information
+        user = user_info_from_headers(key=self.authzKey)
```

Comment:
Isn't key here supposed to be the HTTP header hmac? I am afraid I do not understand the use of `self.authzKey` here.

Reply:
It is defined here https://github.com/dmwm/WMCore/blob/master/src/python/WMCore/REST/Auth.py#L17 and if I read code right it is hmac which should be provided via configuration parameter since it is not known a-priori and we only know it based on FE installation.